kuma/core/middleware.py

```diff
@@ -8,82 +7,204 @@
+
+        if fixed_locale:
+            # Replace the 404 with a redirect to the fixed locale
+            fixed_url = "%s://%s%s" % (
+                request.scheme,
+                request.get_host(),
+                request.get_full_path().replace(literal_from_path,
+                                                fixed_locale,
+                                                1)
+            )
+            redirect_response = HttpResponseRedirect(fixed_url)
+            add_shared_cache_control(redirect_response)
+            return redirect_response
+        else:
+            # No language fixup found, return the 404
             return response
 
-        request.path_info = '/' + prefixer.shortened_path
-        request.LANGUAGE_CODE = prefixer.locale or settings.LANGUAGE_CODE
-        # prefixer.locale can be '', but we need a real locale code to activate
-        # otherwise the request uses the previously handled request's
-        # translations.
-        translation.activate(prefixer.locale or settings.LANGUAGE_CODE)
 
-    def process_response(self, request, response):
-        """Unset the thread-local var we set during `process_request`."""
-        # This makes mistaken tests (that should use LocalizingClient but
-        # use Client instead) fail loudly and reliably. Otherwise, the set
-        # prefixer bleeds from one test to the next, making tests
-        # order-dependent and causing hard-to-track failures.
-        set_url_prefixer(None)
+class LocaleMiddleware(DjangoLocaleMiddleware):
+    """
+    Override the base Django LocaleMiddleware with custom processing for Kuma.
+
+    Differences:
+    * Skip locale detection for known non-localized paths
+    * Use Kuma language codes (en-US) instead of Django's (en-us)
+    * Don't include "Vary: Accept-Language" header
+    * Add caching headers to locale redirects
+    """
+
+    def process_request(self, request):
+        """
+        Determine the language code for the request.
+
+        Differences:
+        * Skip language detection (which adds a Vary header) if the path
+          is a known non-localized path.
+        * Use Kuma language codes (en-US) not Django (en-us)
+        """
+        # If the path is a known non-locale path, skip checks
+        if not self.is_locale_path(request):
+            request.LANGUAGE_CODE = settings.LANGUAGE_CODE
+            return
+
+        # Call Django's LocaleMiddleware to detect the language, activate it,
+        # and set request.LANGUAGE_CODE, based on Django's language code
+        response = super(LocaleMiddleware, self).process_request(request)
+
+        # Replace Django's language code w/ Kuma's, if needed
+        dj_language_code = request.LANGUAGE_CODE
+        kuma_language_code = django_language_code_to_kuma(dj_language_code)
+        if dj_language_code != kuma_language_code:
+            translation.activate(kuma_language_code)
+            request.LANGUAGE_CODE = kuma_language_code
```

Comment:
This whole chunk, lines 151-156, doesn't seem necessary to me since the underlying `django.utils.translation.get_language_from_request` that `DjangoLocaleMiddleware` uses will always return the Kuma language code (for example, the request path `/zh-cn/...` returns `zh-CN` as the language code). It seems to me that this is the case because the underlying `django.utils.translation.get_supported_language_variant` ends-up always called with `strict=False` by `django.utils.translation.get_language_from_path`, and that ensures that `zh-cn` picks up the Kuma `zh-CN` via the generic language code. The only odd case that I can see is that `sr-latn` becomes `sr` while `sr-Latn` remains `sr-Latn`. As before though, maybe I'm missing that this will be needed in the future.